src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -228,31 +221,104 @@ KafkaSink::KafkaSink(const Kafka * kafka, const Block & header, ContextPtr conte
     }
 
-    topic = klog::KTopicPtr(rd_kafka_topic_new(producer.get(), kafka->topic().c_str(), nullptr), rd_kafka_topic_destroy);
-    wb = std::make_unique<WriteBufferFromKafka>(topic.get());
+    topic.reset(rd_kafka_topic_new(producer.get(), kafka->topic().c_str(), nullptr));
+    wb = std::make_unique<WriteBufferFromKafkaSink>([this](char * pos, size_t len) { addMessageToBatch(pos, len); });
 
     String data_format = kafka->dataFormat();
     if (data_format.empty())
         data_format = "JSONEachRow";
 
-    /// The callback allows `IRowOutputFormat` based formats produce one Kafka message per row.
-    if (kafka->produceOneMessagePerRow())
+    if (message_key_ast)
+    {
+        message_key_expr = buildExpression(context, header, message_key_ast);
+        const auto & sample_block = message_key_expr->getSampleBlock();
+        /// The last column is the key column, the others are required columns (to be used to calculate the key value).
+        message_key_column_pos = message_key_expr->getResultPositions().back();
+        auto message_key_column_name = sample_block.getColumnsWithTypeAndName().back().name;
+        /// If the key column already exists in the head, which means the message key is
+        /// one of the column in the stream, in this case, it should not delete the column.
+        /// Otherwise, it's a computed column, it should be deleted from the block.
+        delete_message_key_column = !header.tryGetPositionByName(message_key_column_name).has_value();
+    }
+
+    if (one_message_per_row)
+    {
+        /// The callback allows `IRowOutputFormat` based formats produce one Kafka message per row.
         writer = FormatFactory::instance().getOutputFormat(
             data_format, *wb, header, context, [this](auto & /*column*/, auto /*row*/) { wb->next(); });
+        if (!static_cast<IRowOutputFormat*>(writer.get()))
+            throw Exception(ErrorCodes::INVALID_SETTING_VALUE, "Data format `{}` is not a row-based foramt, it cannot be used with `one_message_per_row`", data_format);
+    }
     else
-        writer = FormatFactory::instance().getOutputFormat(data_format, *wb, header, context);
+    {
+        writer = FormatFactory::instance().getOutputFormat(data_format, *wb, header, context, {});
+    }
     writer->setAutoFlush();
 
     if (kafka->hasCustomShardingExpr())
         partitioner = std::make_unique<KafkaStream::ChunkSharder>(context, header, kafka->shardingExprAst());
     else
         partitioner = std::make_unique<KafkaStream::ChunkSharder>();
 
+    /// Polling message deliveries.
     polling_threads.scheduleOrThrowOnError([this]() {
         while (!is_finished.test())
             if (auto n = rd_kafka_poll(producer.get(), POLL_TIMEOUT_MS))
                 LOG_TRACE(log, "polled {} events", n);
     });
+
+    /// Monitor partition changes.
+    metadata_threads.scheduleOrThrowOnError([this, topic_refresh_interval_ms]() {
```

Comment:
I was thinking we shall merge `metadata_threads / polling message deliveries threads` in one